yt/yt/core/http/retriable_client.cpp

```diff
@@ -0,0 +1,210 @@
+        const TString& url,
+        const TSharedRef& body,
+        const THeadersPtr& headers) override
+    {
+        return MakeRequest(&IClient::Put, responseChecker, url, body, headers);
+    }
+
+    TFuture<IResponsePtr> Delete(
+        const IResponseCheckerPtr& responseChecker,
+        const TString& url,
+        const THeadersPtr& headers) override
+    {
+        return MakeRequest(&IClient::Delete, responseChecker, url, headers);
+    }
+
+private:
+    const TRetrialbeClientConfigPtr Config_;
+    const IInvokerPtr Invoker_;
+
+    IClientPtr UnderlyingClient_;
+
+private:
+    template <typename TCallable, typename... Args>
+    TFuture<IResponsePtr> MakeRequest(
+        TCallable&& func,
+        const IResponseCheckerPtr& responseChecker,
+        const TString& url,
+        Args&&... args)
+    {
+        return BIND([=, this, this_ = MakeStrong(this), func = std::move(func), ...args = std::move(args)] () {
+            const auto deadline = TInstant::Now() + Config_->RequestTimeout;
+            const auto sanitizedUrl = SanitizeUrl(url);
+
+            YT_LOG_DEBUG("Making request (Url: %v, Deadline: %v, Max attempts: %v", sanitizedUrl, deadline, Config_->MaxAttemptCount);
+            std::vector<TError> accumulatedErrors;
+
+            for (int attempt = 1; (TInstant::Now() < deadline && attempt <= Config_->MaxAttemptCount) || attempt == 1; ++attempt) {
+                auto future = BIND(func, UnderlyingClient_, url, std::forward<Args>(args)...)();
+                auto rspOrError = WaitFor(future.WithTimeout(Config_->AttemptTimeout));
+                if (!rspOrError.IsOK()) {
+                    auto error = TError("Request attempt %v failed", attempt)
+                        << rspOrError
+                        << TErrorAttribute("attempt", attempt);
+                    
+                    YT_LOG_WARNING(error, "Request attempt failed (Url: %v, Attempt: %v)", sanitizedUrl, attempt);
+                    accumulatedErrors.push_back(std::move(error));
+                    continue;
+                }
+
+                auto& rsp = rspOrError.Value();
+                const auto responseCheck = responseChecker->CheckError(rsp);
+                if (responseCheck.IsOK()) {
+                    return MakeFuture(rsp);
+                }
+
+                auto error = TError("Error checking response")
+                    << responseCheck
+                    << TErrorAttribute("attempt", attempt);
+                YT_LOG_WARNING(error, "Request attempt failed while checking response (Url: %v, Attempt: %v)", sanitizedUrl, attempt);
+                accumulatedErrors.push_back(std::move(error));
+
+                auto now = TInstant::Now();
+                if (now > deadline) {
+                    break;
+                }
+                TDelayedExecutor::WaitForDuration(std::min(Config_->BackoffTimeout, deadline - now));
+            }
+
+            THROW_ERROR_EXCEPTION("HTTP retriable call failed")
```

Comment:
I think it is better to say simpler: `HTTP request failed` and put number of retries in the error attribute